docs(readme): Add library motivations and goals (#69)
zharinov/parser-utils: README.md

# Renovate parser utils

Code parsing library filling the gap between ad-hoc regular expressions and parsers generated with complete grammar description.

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://github.com/renovatebot/parser-utils/blob/main/LICENSE)
[![Trunk](https://github.com/renovatebot/parser-utils/actions/workflows/trunk.yml/badge.svg)](https://github.com/renovatebot/parser-utils/actions/workflows/trunk.yml)

## Motivation

We are creating Renovate as a multi-language tool for keeping dependency versions up to date.

> While particular package managers are usually making use of their programming language, we needed some uniform way to deal with the variety of dependency description conventions.

Reviewer: ```suggestion
Most package managers use their own programming language, but we still need some uniform way to deal with the variety of dependency description conventions.
```

Author: ```suggestion
While most package managers benefit from relying on the programming language they're written, we needed some uniform way to deal with the variety of dependency description conventions using only TypeScript.
```

> Some package managers just use relatively simple JSON format, like Node ecosystem does.

Reviewer: ```suggestion
Some package managers use the relatively simple JSON format, like Node.js for example.
```

> Others tools, mostly related to CI/CD world, stick to more elaborate YAML.

Reviewer: ```suggestion
Other tools, mostly related to Devops, use the more elaborate YAML format.
```

Author: Maybe, "mostly related to DevOps"?

> But the hardest part is to deal with dependencies described in particular programming languages: for example, Gemfile and Podfile are written in Ruby, `build.gradle` uses Groovy, `sbt` leverages Scala, while `bazel` have created its own language Starlark.

Reviewer: ```suggestion
The trickiest thing is to deal with dependencies described by particular programming languages: for example, Gemfiles and Podfiles are written in Ruby, `build.gradle` files use Groovy, `sbt` leverages Scala, while `bazel` created its own language Starlark.
```

One approach is to use regular expressions, which is very easy but doesn't scale well to cover all syntactic variations.
For example, we want to treat string literals `'foobar'`, `"foobar"` and `"""foobar"""` as equivalent.

> With another approach, we may describe languages with tools like [PEG.js](https://github.com/pegjs/pegjs) or [nearley.js](https://github.com/kach/nearley).

Reviewer: ```suggestion
Another approach could be that we describe languages with tools like [PEG.js](https://github.com/pegjs/pegjs) or [nearley.js](https://github.com/kach/nearley).
```

> Although these are great tools, the approach itself has downsides for our use-case:

Reviewer: ```suggestion
Although these are great tools, this approach has downsides for our use-case:
```

> - We have to define and test complete grammars for each language, even if we're interested mostly in string literals, variable definitions and their scopes.

Reviewer: ```suggestion
- We have to define and test the complete grammar for each language, even if we're interested mostly in string literals, variable definitions and their scopes
```

> - Even small source errors will lead to rejecting the whole file, while we want to just skip the fragments being misunderstood by parser.

Reviewer: ```suggestion
- Even small source errors lead to rejecting the whole file, while we want to skip the fragments that are misunderstood by the parser
```

> - We still would need to deal with variety of language-specific AST tree formats, which may or may not share some common specifics.

Reviewer: ```suggestion
- We still would need to deal with a variety of language-specific AST tree formats, which may or may have things in common
```

Author: Do you mean "may or may not have things in common"?

Reviewer: Whoops! You're right, I meant `may or may not` ... 😄

Reviewer: ```suggestion
- We still would need to deal with a variety of language-specific AST tree formats, which may or may not have things in common.
```

> This library is an attempt to fill the gap between these two approaches.

Reviewer: ```suggestion
The `parser-utils` library is an attempt to fill the gap between these two approaches.
```

> It leverages [moo](https://github.com/no-context/moo) tokenizer library, organizes tokens into trees and helps to query them.

Reviewer: ```suggestion
It leverages the [moo](https://github.com/no-context/moo) tokenizer library, organizes tokens into trees and helps to query them.
```

> The query API is inspired by [parsimmon](https://github.com/jneen/parsimmon), though it operates on parsed tokens instead of raw source code.

Reviewer: ```suggestion
The query API is inspired by [parsimmon](https://github.com/jneen/parsimmon), though it operates on parsed tokens instead of the raw source code.
```

## Goals

- _Be good enough for source code written well enough._

> - Go much further than it's possible with regular expressions.

Reviewer: ```suggestion
- Go much further than is possible with regular expressions.
```

> - Respect location info. Once something interesting is found, it can be located in source test via offset info. Once something is written, it should not affect whole document formatting.

Reviewer: ```suggestion
- Respect location info. Once something interesting is found, it can be located in the source test via offset info. Once something is written, it should not affect the whole document formatting.
```

> - Incorporate poorly recognized fragments into the output and just continue parsing.

Reviewer: ```suggestion
- Incorporate poorly recognized fragments into the output and continue parsing.
```

> - Expressive API helping to focus on syntactic structure, not the space or quote styles. When details are necessary, there is a way to specify them too.

Reviewer: ```suggestion
- Expressive API which helps users to focus on the syntactic structure, not the space or quote styles. When details are necessary, they can be specified.
```

Author: Maybe, `explicitly specified` or `specified explicitly`?

Reviewer: ```suggestion
- Expressive API which helps you focus on syntactic structure, not the space or quote variations.
```

Maybe something like this would be better? Please adjust the suggestion before you apply it though, I'm not sure _what_ things you can specify in more detail, so you'll need to add that in your yourself. 😉

> - Allow to define a language of interest quickly. Provide definitions for popular languages out of the box.

Reviewer: ```suggestion
- Allow to define a language of interest quickly. Provide definitions for popular languages out-of-the-box.
```
